hathor/transaction/static_metadata.py

```diff
@@ -245,3 +248,47 @@ def _calculate_my_min_height(
             if isinstance(spent_tx, Block):
                 min_height = max(min_height, spent_tx.static_metadata.height + settings.REWARD_SPEND_MIN_BLOCKS + 1)
         return min_height
+
+    @staticmethod
+    def _calculate_closest_ancestor_block(
+        tx: 'Transaction',
+        settings: HathorSettings,
+        vertex_getter: Callable[[VertexId], 'BaseTransaction'],
+    ) -> VertexId:
+        """
+        Calculate the tx's closest_ancestor_block. It's calculated by propagating the metadata forward in the DAG.
+        """
+        from hathor.transaction import Block, Transaction
+        if tx.is_genesis:
+            return settings.GENESIS_BLOCK_HASH
+
+        closest_ancestor_block: Block | None = None
+
+        for vertex_id in tx.get_all_dependencies():
+            vertex = vertex_getter(vertex_id)
+            candidate_block: Block
+
+            if isinstance(vertex, Block):
+                candidate_block = vertex
+            elif isinstance(vertex, Transaction):
+                vertex_candidate = vertex_getter(vertex.static_metadata.closest_ancestor_block)
+                assert isinstance(vertex_candidate, Block)
+                candidate_block = vertex_candidate
+            else:
+                raise NotImplementedError
+
+            if (
+                not closest_ancestor_block
+                or candidate_block.static_metadata.height > closest_ancestor_block.static_metadata.height
+            ):
+                closest_ancestor_block = candidate_block
```

Comment:
No, the `closest_ancestor_block` is just a DAG property, much like `parents` and `inputs` are. In fact, it is literally just a derivation based on `parents` and `inputs`.

Considering its use for Feature Activation, voided blocks carry their feature states normally just like non-voided blocks do. For transactions this will be analogous: the feature states of the transaction will be the same of its `closest_ancestor_block`, even if it's voided. Naturally, the transaction will be voided too, since it depends on this block.

Reply:
Thanks for the clarification.